advanced_source/cpp_cuda_graphs/mnist.cpp

```diff
@@ -0,0 +1,378 @@
+  model.train();
+
+  at::cuda::CUDAStream warmupStream = at::cuda::getStreamFromPool();
+  at::cuda::CUDAStream captureStream = at::cuda::getStreamFromPool();
+  at::cuda::CUDAStream legacyStream = at::cuda::getCurrentCUDAStream();
+
+  at::cuda::setCurrentCUDAStream(warmupStream);
+
+  stream_sync(legacyStream, warmupStream);
+
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    training_step(model, optimizer, data, targets, output, loss);
+  }
+
+  stream_sync(warmupStream, captureStream);
+  at::cuda::setCurrentCUDAStream(captureStream);
+
+  graph.capture_begin();
+  training_step(model, optimizer, data, targets, output, loss);
+  graph.capture_end();
+
+  stream_sync(captureStream, legacyStream);
+  at::cuda::setCurrentCUDAStream(legacyStream);
+}
+
+template <typename DataLoader>
+void train(
+    size_t epoch,
+    Net& model,
+    torch::Device device,
+    DataLoader& data_loader,
+    torch::optim::Optimizer& optimizer,
+    size_t dataset_size,
+    torch::Tensor& data,
+    torch::Tensor& targets,
+    torch::Tensor& output,
+    torch::Tensor& loss,
+    at::cuda::CUDAGraph& graph,
+    bool use_graph) {
+  model.train();
+
+  size_t batch_idx = 0;
+
+  for (auto& batch : data_loader) {
+    if (batch.data.size(0) != kTrainBatchSize ||
+        batch.target.size(0) != kTrainBatchSize) {
+      continue;
+    }
+
+    data.copy_(batch.data);
+    targets.copy_(batch.target);
+
+    if (use_graph) {
+      graph.replay();
+    } else {
+      training_step(model, optimizer, data, targets, output, loss);
+    }
+
+    if (batch_idx++ % kLogInterval == 0) {
+      float train_loss = loss.item<float>();
+      std::printf(
+          "\rTrain Epoch: %ld [%5ld/%5ld] Loss: %.4f",
+          epoch,
+          batch_idx * batch.data.size(0),
+          dataset_size,
+          train_loss);
+    }
+  }
+}
+
+void test_step(
+    Net& model,
+    torch::Tensor& data,
+    torch::Tensor& targets,
+    torch::Tensor& output,
+    torch::Tensor& loss) {
+  output = model.forward(data);
+  loss = torch::nll_loss(output, targets, {}, torch::Reduction::Sum);
+}
+
+void capture_test_graph(
+    Net& model,
+    torch::Tensor& data,
+    torch::Tensor& targets,
+    torch::Tensor& output,
+    torch::Tensor& loss,
+    torch::Tensor& total_loss,
+    torch::Tensor& total_correct,
+    at::cuda::CUDAGraph& graph) {
+  torch::NoGradGuard no_grad;
+  model.eval();
+
+  const int num_warmup_iters = 7;
+
+  at::cuda::CUDAStream warmupStream = at::cuda::getStreamFromPool();
+  at::cuda::CUDAStream captureStream = at::cuda::getStreamFromPool();
+  at::cuda::CUDAStream legacyStream = at::cuda::getCurrentCUDAStream();
+
+  at::cuda::setCurrentCUDAStream(warmupStream);
+  stream_sync(captureStream, legacyStream);
+
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    test_step(model, data, targets, output, loss);
+    total_loss = total_loss + loss;
+    total_correct = total_correct + output.argmax(1).eq(targets).sum();
```

Comment:
Why not use `+=` here?
```suggestion
    total_loss += loss;
    total_correct += output.argmax(1).eq(targets).sum();
```